File: src/token.rs


pub struct Token {
    pub value: String,
    pub token_type: TokenType,
    pub line: i32
}

pub enum TokenType {
    // Single-character tokens.
    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    Comma,
    Dot,
    Minus,
    Plus,
    Semicolon,
    Slash,
    Star,

    // One or two character tokens.
    Bang,
    BangEqual,
    Equal,
    EqualEqual,
    Greater,
    GreaterEqual,
    Lesser,
    LesserEqual,

    // Literals.
    Identifier,
    String,
    Number,

    // Keywords.
    And,
    Class,
    Else,
    False,
    Func,
    For,
    If,
    Null,
    Or,
    Print,
    Return,
    Super,
    This,
    True,
    Var,
    While,

    Eof
}

File: token/token.go
package token

const (
	// LeftParen : (
	LeftParen = iota
	// RightParen : )
	RightParen
	// LeftBrace : {
	LeftBrace
	// RightBrace : }
	RightBrace
	// Comma : ,
	Comma
	// Dot : .
	Dot
	// Minus : -
	Minus
	// Plus : +
	Plus
	// Semicolon : ;
	Semicolon
	// Slash : /
	Slash
	// Star : *
	Star

	// Bang : !
	Bang
	// BangEqual : !=
	BangEqual
	// Equal : =
	Equal
	// EqualEqual : ==
	EqualEqual
	// Greater : >
	Greater
	// GreaterEqual : >=
	GreaterEqual
	// Less : <
	Less
	// LessEqual : <=
	LessEqual

	// Identifier : Any variable
	Identifier
	// String : String Literal
	String
	// Number : Number Literal
	Number

	// And : &
	And
	// Class : class Keyword
	Class
	// Else : else Keyword
	Else
	// False : false keyword
	False
	// Func : func keyword
	Func
	// For : for keyword
	For
	// If : if keyword
	If
	// Nil : nil keyword
	Nil
	// Or : or keyword
	Or
	// Print : print keyword
	Print
	// Return : return keyword
	Return
	// Super : super keyword
	Super
	// This : this keyword
	This
	// True : true literal
	True
	// Var : var keyword
	Var
	// While : while keyword
	While
	// EOF : end of file indicator
	EOF
)

// Type for token types.
// TokenType denotes the type of the token. It is to by pass the enum constrain in go.
type Type int

// Token is structure for token details.
type Token struct {
	TokenType Type
	Lexeme    string
	Line      int
}
